src/puppetlabs/puppetdb/queue.clj

```diff
@@ -0,0 +1,199 @@
+(ns puppetlabs.puppetdb.queue
+  (:import [java.nio.charset StandardCharsets]
+           [java.io InputStreamReader BufferedReader InputStream]
+           [java.util TreeMap HashMap]
+           [java.nio.file Files LinkOption]
+           [java.nio.file.attribute FileAttribute])
+  (:require [clojure.string :as str :refer [re-quote-replacement]]
+            [puppetlabs.stockpile.queue :as stock]
+            [clj-time.coerce :as tcoerce]
+            [puppetlabs.puppetdb.cheshire :as json]
+            [puppetlabs.puppetdb.command.constants :as constants]
+            [metrics.timers :refer [timer time!]]
+            [metrics.counters :refer [inc!]]
+            [clj-time.core :as time]
+            [puppetlabs.kitchensink.core :as kitchensink]
+            [slingshot.slingshot :refer [throw+]]
+            [clojure.core.async :as async]
+            [clojure.core.async.impl.protocols :as async-protos]
+            [puppetlabs.puppetdb.nio :refer [get-path]]))
+
+(def metadata-command-names
+  (vals constants/command-names))
+
+(defn stream->json [^InputStream stream]
+  (try
+    (-> stream
+        (InputStreamReader. StandardCharsets/UTF_8)
+        BufferedReader.
+        (json/parse-stream true))
+    (catch Exception e
+      (throw+ {:kind ::parse-error} e "Error parsing command"))))
+
+(defn metadata-str [received command version certname]
+  (format "%d_%s_%d_%s.json"
+          (tcoerce/to-long received) command version certname))
+
+(defn- metadata-rx [valid-commands]
+  (re-pattern (str
+               "([0-9]+)_("
+               (str/join "|" (map #(format "(?:%s)" (re-quote-replacement %))
+                                  valid-commands))
+               ")_([0-9]+)_(.*)\\.json")))
+
+(defn metadata-parser
+  ([] (metadata-parser metadata-command-names))
+  ([valid-commands]
+   ;; NOTE: changes here may affect the DLO, e.g. it currently assumes
+   ;; the trailing .json.
+   (let [rx (metadata-rx valid-commands)]
+     (fn [s]
+       (when-let [[_ stamp command version certname] (re-matches rx s)]
+         (and certname
+              {:received (-> stamp
+                             Long/parseLong
+                             tcoerce/from-long
+                             kitchensink/timestamp)
+               :version (Long/parseLong version)
+               :command command
+               :certname certname}))))))
+
+(def parse-metadata (metadata-parser))
+
+(def parse-cmd-filename
+  (let [parse-metadata (metadata-parser (cons "unknown" metadata-command-names))]
+    (fn [s]
+      (let [rx #"([0-9]+)-(.*)"]
+        (when-let [[_ id qmeta] (re-matches rx s)]
+          (parse-metadata qmeta))))))
+
+(defrecord CommandRef [id command version certname received callback annotations delete?])
```

Comment:
This is the "ticket stub" that PuppetDB uses to dequeue messages. This needs to be as lean as possible as there could be many of them in memory. When we switch from the proper command names to a short hand when writing the filenames to disk, we should be able to intern command, version is a number along with received.